### DISCUSSION TITLE
How to prevent a TextBox from losing focus when clicking a Button

### DISCUSSION THREAD

**Discussion by msneijders** [AvaloniaUI/Avalonia#16443] [Q&A] [answered]
For on a touch (android) device) I would like a NumericUpDown control that has this behavior:
- Tap on the textbox to give it focus
- On-screen keyboard pops up for input.
- I click a spinner button to increase the value.
- The input focus must stay on the textbox, so on-screen keyboard stays visible.

(if the textbox did not have focus when clicking on the button, it should _not_ get focus -- focus should simply not change)

I tried this, using `Focusable=False`:

```csharp
<StackPanel>
   <TextBox Text="1234 />
   <Button Focusable="False">Clickto increase value but no change in focus</Button>
</StackPanel>
```

Although this does not give focus to the button, it moves to focus to a parent element -- the textbox loses focus.

Does anyone know how to not do anything with focus when you click/tap a button?

**Comment by IanRawley** [comment]
What you want looks to be non-trivial based on how early FocusManager gets its mitts on PointerPressed events.  My best guess for a work-around would be to make the button(s) children of the Textbox and non-focusable, either by inserting them in InnerLeftContent or InnerRightContent, or creating an entirely new template for Textbox which includes the Buttons so that they are effectively one control.  Either way I think you'll still see the onscreen keyboard disappear and reapppear when the buttons are pressed unless you go whole-hog into making an entirely new control.

**Reply by msneijders** [reply]
My example is a bit misleading: but actually it already is one control, the built-in NumericUpDown. Which is a TextBox inside a ButtonSpinner. And although the buttons themselves will not get focus when setting Focusable=False, it is then the ButtonSpinner that gets focus, the TextBox will lose focus and on-screen keyboard disappears.

The `FocusManager` does not leave much, or any, room for other behavior it seems.

ty for your reponse.

**Reply by IanRawley** [reply]
I guessed as much.  But as you said that's a Textbox inside a ButtonSpinner.  I mean make it a ButtonSpinner inside a TextBox, so as to exploit the FocusManager focusing the parent if the target can't be focused.  If the Button can't be focused and it's a child of the TextBox, the TextBox gains focus.   I expect the OSK still disappears and reappears.

Alternatively I mean make it literally one control.  No actual button, just something that looks like a button (like a heavily styled TextBlock) but is actually part of the "TextBox".  But that's a lot of work.

**Comment by workgroupengineering** [comment, marked as answer]
A partial workauroud is to add the following style selector

```xaml
        <Style Selector=":is(NumericUpDown)">
            <Setter Property="Focusable" Value="False"/>
            <Style Selector="^/template/ ButtonSpinner">
                <Setter Property="Focusable" Value="False"/>
                <Style Selector="^/template/ RepeatButton">
                    <Setter Property="Focusable" Value="False"/>
                </Style>
            </Style>
        </Style>
```

**Reply by msneijders** [reply]
Thank you; got it working -- I didn't put Focusable=false on the control (NumericUpDown) itself -- I assumed that would prevent it from getting focus while using tab or otherwise a "focus next" -- but its child element TextBox does get focus.
